### ISSUE TITLE
find not working correctly for regular expression containing asterisk wildcard

### ISSUE THREAD

**Issue by overtomanu** [microsoft/vscode#116445]
I am trying to find with regex "[^\t]*" in a tab delimited file

paste the following contents in vs code in untitled buffer where fields are delimited by tabs
f1	f2
f3	f4
f5	f6

then try to find in regex mode with regular expression "[^\t]*"
observe that it finds only f1,f3 and f5 and completely ignores fields in second column. *  means it should match zero of more characters, so it should match fields in second column.

if I try to find with regex "[^\t]+" it works correctly and does find fields in second column.

details about the VS code build I am using:
Version: 1.53.1
Commit: 5d424b828ada08e1eb9f95d6cb41120234ef57c7
Date: 2021-02-08T23:35:15.302Z (2 days ago)
Electron: 11.2.1
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Darwin x64 19.6.0

**Comment by IllusionMH** [comment]
It finds and highlights not only values from column (Press <kbd>Alt</kbd>+<kbd>Enter</kbd> to select all matches and past somewhere to see how everything) but also some empty string after first column and before `\t`.

However looks like these zero-length matches make it "jump over" all other matches in that line.
![image](https://user-images.githubusercontent.com/1312662/107632681-7b419f80-6c6f-11eb-8e0a-756e733dcde6.png)

**Comment by overtomanu** [comment]
Yes that's my observation as well. I think if there is a empty line then it should match empty string (because * means zero or more. This is what happens in other editors like notepad++.)
But it should not match empty string just by the side of a matched string. That's just wrong.

**Comment by IllusionMH** [comment]
That's JS behavior which you can observe in browser console looking at using `[...'f1\tf2\tf3\tf4'.matchAll(/[^\t]*/g)]`.
So "That's just wrong." should got to TC39 committee I guess 😄 

It will really depend on RE engine. Looks like ripgrep (used for Find in files when you don't have file opened) will match empty string at the end of each string instead.

Skipping zero-length matches at the end looks like something that could help in this case, however IMHO it's not always "valid" to always discard empty matches because it's regexps - you can write regexp that will produce valid zero length in different "branches" in consecutive form.

**Comment by ArturoDent** [comment]
Also, regex101, see https://regex101.com/r/BdW56f/1, and regexr, see https://regexr.com/5map5, match those empty strings as well.  I think that is correct behaviour.

What is odd is how using the find widgets up/down arrows work - that seems odd - first you arrow down one side, but if you arrow down and then arrow up you see it does go to each match correctly (in some odd order).  But if you <kbd>Ctrl</kbd>+<kbd>shift</kbd>+<kbd>L</kbd> to select all the find matches you see it did find them all.  It is just the order of cycling through the find matches has some "logic" I don't understand.  It would seem that f2 would be the next find match after f1 and then the empty string right after f1, but the arrows don't go in that order.

**Comment by roblourens** [comment]
Thanks @IllusionMH. Don't know what to say about where the regex engine decides to put empty matches, but this is weird and is something we do have control over:

![Feb-12-2021 09-38-57](https://user-images.githubusercontent.com/323878/107802425-286ef300-6d16-11eb-8592-ec429590704f.gif)

**Comment by IllusionMH** [comment]
@roblourens I'm not arguing that Next should go to next match instead of skipping whole line and I'm looking into it at this moment.

I'm just not sure that it's valid to filter out zero-length matches in this case. However I've checked parts with navigation and looks like they should "skip" skip empty matches in this case just don't do it properly.

What would be expected solution here - remove these zero-length matches from results in this case or just properly on Next?
Removing matches here might also require to filter out matches in Find in Files as well.

**Comment by roblourens** [comment]
I don't think we should filter out any matches here

**Comment by IllusionMH** [comment]
@roblourens empty matches already skipped when user clicks on Next match

https://github.com/microsoft/vscode/blob/6c12d9f2c4aa9f14ce1ca81033ebf31ea48c0bc0/src/vs/editor/contrib/find/findModel.ts#L429
And wrong position from this call will skip all all results in same line which is root cause of this issue
https://github.com/microsoft/vscode/blob/6c12d9f2c4aa9f14ce1ca81033ebf31ea48c0bc0/src/vs/editor/contrib/find/findModel.ts#L431

I will soon make separate PRs that
1) Fix incorrect position returned from `_nextSearchPosition`
2) Removes empty ranges that are at the end of previous matches (anyway they will be skipped as shown above and user when click Next will see 1 of 9, 3 of 9, 4 of 9, 6 of 6 etc.)

to continue discussion with working examples.

**Comment by roblourens** [comment]
Oh, thanks for investigating. I guess I would wait to hear from @rebornix before you spend a bunch of time on this, since he actually owns this feature.

**Comment by IllusionMH** [comment]
My results so far for both changes approved (but I know that none might be)

https://user-images.githubusercontent.com/1312662/108047916-0bf0f480-704f-11eb-86d9-ebbf2a59e414.mp4
